google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITSystemTest.java

```diff
@@ -1686,4 +1688,110 @@ public void testBuildingBundleWithLimitToLastQuery() throws Exception {
+    for (DocumentReference documentReference : documents) {
+      count += countDocumentChildren(documentReference) + 1;
+    }
+    return count;
+  }
+
+  private void setupRecursiveDeleteTest() throws Exception {
+    // ROOT-DB
+    // └── randomCol
+    //     ├── anna
+    //     └── bob
+    //         └── parentsCol
+    //             ├── charlie
+    //             └── daniel
+    //                 └── childCol
+    //                     ├── ernie
+    //                     └── francis
+    WriteBatch batch = firestore.batch();
+    batch.set(randomColl.document("anna"), map("name", "anna"));
+    batch.set(randomColl.document("bob"), map("name", "bob"));
+    batch.set(randomColl.document("bob/parentsCol/charlie"), map("name", "charlie"));
+    batch.set(randomColl.document("bob/parentsCol/daniel"), map("name", "daniel"));
+    batch.set(randomColl.document("bob/parentsCol/daniel/childCol/ernie"), map("name", "ernie"));
+    batch.set(
+        randomColl.document("bob/parentsCol/daniel/childCol/francis"), map("name", "francis"));
+    batch.commit().get();
+  }
+
+  @Test
+  public void testRecursiveDeleteTopLevelCollection() throws Exception {
+    setupRecursiveDeleteTest();
+    firestore.recursiveDelete(randomColl).get();
+    assertEquals(0, countCollectionChildren(randomColl));
+  }
+
+  @Test
+  public void testRecursiveDeleteNestedCollection() throws Exception {
+    setupRecursiveDeleteTest();
+    firestore.recursiveDelete(randomColl.document("bob").collection("parentsCol")).get();
+    assertEquals(2, countCollectionChildren(randomColl));
+  }
+
+  @Test
+  public void testRecursiveDeleteNestedDocument() throws Exception {
+    setupRecursiveDeleteTest();
+    DocumentReference document = randomColl.document("bob/parentsCol/daniel");
+    firestore.recursiveDelete(document).get();
+    DocumentSnapshot snap = document.get().get();
+    assertFalse(snap.exists());
+    assertEquals(1, countDocumentChildren(randomColl.document("bob")));
+    assertEquals(3, countCollectionChildren(randomColl));
+  }
+
+  @Test
+  public void testRecursiveDeleteLeafDocument() throws Exception {
+    setupRecursiveDeleteTest();
+    DocumentReference document = randomColl.document("bob/parentsCol/daniel/childCol/ernie");
+    firestore.recursiveDelete(document).get();
+    DocumentSnapshot snap = document.get().get();
+    assertFalse(snap.exists());
+    assertEquals(5, countCollectionChildren(randomColl));
+  }
+
+  @Test
+  public void testRecursiveDeleteDoesNotAffectOtherCollections() throws Exception {
+    setupRecursiveDeleteTest();
+
+    // Add another nested collection that shouldn't be deleted.
+    CollectionReference collectionB = firestore.collection("doggos");
+    collectionB.document("doggo").set(map("name", "goodboi"));
+
+    firestore.recursiveDelete(collectionB).get();
+    assertEquals(6, countCollectionChildren(randomColl));
+    assertEquals(0, countCollectionChildren(collectionB));
```

Comment:
Should there not still be one document?

Reply:
Got it. I am glad you made me look at this again though - I think you should wait for "doggo" to be written before invoking the delete.